src/extra/components/detector_motors.py

```diff
@@ -0,0 +1,414 @@
+                If any motor source is not found
+            PropertyNameError:
+                If any motor position key is not found
+        """
+        self.detector_id = detector_id
+        self.src_ptrn = src_ptrn
+        self.key_ptrn = key_ptrn
+        self.dims = dims
+
+        names, coordinates = zip(*self.dims.items())
+        self.motor_ids = [(
+                src_ptrn.format(**dict(zip(names, values))),
+                key_ptrn.format(**dict(zip(names, values)))
+            ) for values in product(*coordinates)]
+        self.motor_labels = [
+            ''.join(f"{n.upper()}{v}" for n, v in zip(names, values))
+            for values in product(*coordinates)]
+
+        self.shape = tuple(len(c) for c in coordinates)
+        self.num_sources = np.prod(self.shape)
+        self.dc = dc.select(self.motor_ids)
+
+        self.num_trains = len(self.dc.train_ids)
+
+        # check sources
+        self.keys = []
+        for src, key in self.motor_ids:
+            self.keys.append(self.dc[src, key])
+
+        # add train_ids
+        self.train_ids = np.arange(
+            self.dc.train_ids[0], self.dc.train_ids[-1] + np.uint64(1),
+            dtype=np.uint64)
+
+    def positions(self, labelled=True):
+        """Returns the motor positions for all trains.
+
+        Args:
+            labelled (bool):
+                If True, returns the xarray with labelled dimensions,
+                overwise returns numpy.ndarray
+
+        Returns:
+            positions (numpy.ndarray or xarray.DataArray):
+                The motor positions
+        """
+        train_ids, pos, counts = self._read_positions()
+        pos = np.repeat(pos, counts, axis=0)
+
+        if labelled:
+            dims = ["trainId"] + list(self.dims.keys())
+            coords = {"trainId": self.train_ids}
+            coords.update({name: values for name, values in self.dims.items()})
+            return xarray.DataArray(
+                pos, dims=dims, coords=coords, name=self._position_key)
+        else:
+            return pos
+
+    def _read_positions(self):
+        """Reads and compresses motor positions."""
+        if hasattr(self, "_positions"):
+            return self._positions
+
+        # read positions
+        pos = np.zeros((self.num_trains, self.num_sources), dtype=float)
+        for source_no, key_data in enumerate(self.keys):
+            pos[:, source_no] = key_data.ndarray()
+        pos = pos.reshape(-1, *self.shape)
+
+        # compress
+        axes = tuple(range(1, pos.ndim))
+        ix = np.flatnonzero(
+            np.insert(np.any(np.diff(pos, axis=0), axis=axes), 0, True))
+
+        # train ids when positions are changed
+        train_ids = np.asarray(self.dc.train_ids)[ix]
+        # count of trains at every position
+        counts = np.diff(np.append(train_ids.astype(int),
+                                   int(self.dc.train_ids[-1]) + 1))
+
+        self._positions = train_ids, pos[ix], counts
```

Comment:
I dunno about saving the compressed arrays as an internal representation, it feels unnecessarily complicated. Why not store the entire position arrays in memory?

Reply:
I'm not sure if DAQ creates records for all trains. It should, in principle, but I've decided to not rely on it for protection reason. With compressed version we aren't attached to the stored trains anymore, and can return motor positions for any train between changes.

Compressed version is also allows to see directly if there are changes during the run.

Since this, it makes sense to expose compressed version. If we expose, then why not to store as compressed? The compression and decompression aren't really complicated. They are just four and one line.

Reply:
Ehhh, we definitely shouldn't make up data for trains that aren't saved :sweat_smile: But if `positions_at()` will throw when given a bad train ID then I guess it won't happen anyway.

> Since this, it makes sense to expose compressed version. If we expose, then why not to store as compressed? The compression and decompression aren't really complicated. They are just four and one line.

So one reason would be a simpler internal representation, the other would be a more efficient implementation of `positions_at()` because it wouldn't have to do a search every time (though TBF it's presumably a binary search so it won't cost much anyway). But I don't feel strongly about it so let's continue with it as is.